File: killstreek/src/main/java/dev/kkorolyov/killstreek/item/ArmorSet.java
package dev.kkorolyov.killstreek.item;

import dev.kkorolyov.killstreek.item.Armor.Type;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * A collection of armor pieces equipped to slots.
 */
public class ArmorSet {
	private final Map<Type, Armor> armor = new HashMap<>();

	/** @return total defense value of all unbroken armor */
	public int getDefense() {
		return armor.values().stream()
				.filter(e -> !e.isBroken())
				.mapToInt(Equippable::getValue)
				.sum();
	}

	/** @return equipped armor pieces */
	public Iterable<Armor> getArmor() {
		return armor.values();
	}
	/**
	 * @param broken {@code true} filters to broken armor, {@code false} filters to unbroken armor
	 * @return filtered equipped armor pieces
	 */
	public Iterable<Armor> getArmor(boolean broken) {
		return armor.values().stream()
				.filter(e -> e.isBroken() == broken)
				.collect(Collectors.toList());
	}

	/**
	 * @param armor new equipped armor piece
	 * @return previous equipped armor of the same type as {@code armor}, or {@code null} if no such type was equipped
	 */
	public Armor equip(Armor armor) {
		return this.armor.put(armor.getType(), armor);
	}
	/**
	 * @param type type of armor to unequip
	 * @return unequipped armor, or {@code null} if no such type was equipped
	 */
	public Armor unequip(Armor.Type type) {
		return armor.remove(type);
	}
}

File: killstreek/src/main/kotlin/dev/kkorolyov/killstreek/item/ArmorSet.kt
package dev.kkorolyov.killstreek.item

import dev.kkorolyov.killstreek.item.Armor.Type
import java.util.EnumMap

/**
 * A collection of armor pieces equipped to slots.
 * @constructor constructs a new armor set with initial [armor] pieces
 */
class ArmorSet(armor: Iterable<Armor>) {
	private val _armor: MutableMap<Type, Armor> = EnumMap(Type::class.java)

	/** Current armor pieces */
	val armor: Collection<Armor> get() = _armor.values

	/** total defense value of all unbroken armor */
	val defense: Int = this._armor.values
		.filter { !it.isBroken }
		.sumBy { it.value }

	init {
		armor.forEach { equip(it) }
	}

	/** see [equip] */
	operator fun plusAssign(armor: Armor) {
		equip(armor)
	}

	/** see [unequip] */
	operator fun minusAssign(type: Type) {
		unequip(type)
	}

	/**
	 * Equips [armor] and returns previously equipped piece of the same type.
	 */
	fun equip(armor: Armor): Armor? = _armor.put(armor.type, armor)

	/**
	 * Removes and returns current equipped armor of [type].
	 */
	fun unequip(type: Type): Armor? = _armor.remove(type)
}
